harp/reader.py

```diff
@@ -44,13 +45,34 @@ def __init__(
         self.read = read
 
 
+class RegisterMap(UserDict[str, RegisterReader]):
+    _address_map: dict[int, RegisterReader]
+
+    def __init__(self, registers: dict[str, RegisterReader]) -> None:
+        super().__init__(registers)
+        self._address_map = {
+            value.register.address: value for value in registers.values()
+        }
+
+    def __getitem__(self, __key: Union[str, int]) -> RegisterReader:
+        if isinstance(__key, int):
+            return self._address_map[__key]
+        else:
+            return super().__getitem__(__key)
+
+
 class DeviceReader:
     device: Model
-    registers: dict[str, RegisterReader]
+    registers: RegisterMap
 
-    def __init__(self, device: Model, registers: dict[str, RegisterReader]) -> None:
+    def __init__(
+        self, device: Model, registers: Union[RegisterMap, dict[str, RegisterReader]]
+    ) -> None:
```

Comment:
I think we can leave this code in its original simpler form and create always a new `RegisterMap`.

```suggestion
    def __init__(self, device: Model, registers: dict[str, RegisterReader]) -> None:
```